Issue #8141: Dash validator allowing None values in addition to floats

kenmaca opened the pull request:
Validators for dashed linestyles now allow `None` as an allowed value along with floats if optional `allow_none` kwarg flag in `validate_nseq_float`.

Other validators that use `validate_nseq_float` aren't affected, so this bugfix won't have any breaking changes.

QuLogic commented on lib/matplotlib/rcsetup.py:
```diff
@@ -294,8 +294,8 @@ def validate_maskedarray(v):
-    def __init__(self, n=None):
-        self.n = n
+    def __init__(self, n=None, allow_none=False):
+        self.n, self.allow_none = n, allow_none
```
Is it very common here to put this on one line?

QuLogic commented on lib/matplotlib/rcsetup.py:
```diff
@@ -309,7 +309,11 @@ def __call__(self, s):
-            return [float(val) for val in s]
+            return [float(val)
+                    if self.allow_none and val is not None
+                    or not self.allow_none
```
I think this can be simplified to `not self.allow_none or val is not None`.

QuLogic commented on lib/matplotlib/rcsetup.py:
```diff
@@ -929,7 +933,6 @@ def _validate_linestyle(ls):
     raise ValueError("linestyle must be a string or " +
                      "an even-length sequence of floats.")
 
-
```
Why remove?

QuLogic commented on lib/matplotlib/rcsetup.py:
```diff
@@ -964,7 +967,8 @@ def _validate_linestyle(ls):
     'lines.dashed_pattern':  [[3.7, 1.6], validate_nseq_float()],
-    'lines.dashdot_pattern': [[6.4, 1.6, 1, 1.6], validate_nseq_float()],
+    'lines.dashdot_pattern': [[6.4, 1.6, 1, 1.6],
+                              validate_nseq_float()],
```
Why change this?

kenmaca commented:
Thanks @QuLogic. I apologize for the novice mistakes and have made appropriate corrections. This seems to fix the initial problem, but it seems CI is failing -- will investigate further.

dkua commented:
It would probably be good to include a test for the regression this PR is fixing. There's a good place in https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/tests/test_cycles.py to add in a test. You might also want to think about rebasing the "Fixed" and "Changed" into one commit as it's better PR style.

I'm assuming you're doing this for CSCD01? If you are, tell Anya I said hi.

kenmaca commented:
I've added a test and cleaned up my commits. Looks like AppVeyor decided to play nice this time -- couldn't figure out why it was failing the single build (likely something unrelated).

QuLogic commented:
Looks okay, I think, but can you amend the commit to give the correct first line.

QuLogic commented:
Backported to `v2.0.x` as 058f6b1ee2a6531b2fbfa4e8ac0d25173ccd9be0.
